odoo/addons/base/models/res_users.py

```diff
@@ -478,31 +483,52 @@ def _check_credentials(self, credential, env):
           - { 'uid': 32, 'auth_method': 'webauthn',      'mfa': 'skip'    }
         :rtype: dict
         """
-        if not (credential['type'] == 'password' and credential['password']):
-            raise AccessDenied()
-        self.env.cr.execute(
-            "SELECT COALESCE(password, '') FROM res_users WHERE id=%s",
-            [self.env.user.id]
-        )
-        [hashed] = self.env.cr.fetchone()
-        valid, replacement = self._crypt_context()\
-            .verify_and_update(credential['password'], hashed)
-        if replacement is not None:
-            self._set_encrypted_password(self.env.user.id, replacement)
-            if request and self == self.env.user:
-                self.env.flush_all()
-                self.env.registry.clear_cache()
-                # update session token so the user does not get logged out
-                new_token = self.env.user._compute_session_token(request.session.sid)
-                request.session.session_token = new_token
-
-        if not valid:
-            raise AccessDenied()
-        return {
-            'uid': self.env.user.id,
-            'auth_method': 'password',
-            'mfa': 'default',
-        }
+        env = env or {}
+        interactive = env.get('interactive', True)
+
+        if interactive or not self.env.user._rpc_api_keys_only():
+            if 'interactive' not in env:
+                _logger.warning(
+                    "_check_credentials without 'interactive' env key, assuming interactive login. \
+                    Check calls and overrides to ensure the 'interactive' key is properly set in \
+                    all _check_credentials environments"
+                )
+
+            if not (credential['type'] == 'password' and credential['password']):
```

Comment:
Currently whether you use an API key or a password to authenticate in XMLRPC, it's gonna be a `credential['type'] == 'password'`

https://github.com/odoo/odoo/blob/d10d14e5d8100bcb7bbdcd104f517c0512656795/odoo/service/common.py#L29-L30

But I get your point that this is not clear.

> Maybe the not interactive block should be lifted above this one instead of being in fallback position? And the password version is the fallback?

I was thinking about it, but I wanted to keep the behavior as close as possible as before, as this is supposed to be only a move/merge of code, and no change of behavior.

Also there are also most-likely more people out there using password as xmlrpc authentication rather than api keysm so I was thnking "performance-wise" maybe it's better to keep the regular password first and api key should be the fall back.

What I can maybe do is to not raise in case of `not (credential['type'] == 'password' and credential['password']):` but rather wrap the entire password checking block in `if credential['type'] == 'password' and credential['password']:`

I already changed a bit the logic with the `if valid` anyways

It was
```py
        if not valid:
            raise AccessDenied()
```

Now it is
```py

            if valid:
                return {
                    'uid': self.env.user.id,
                    'auth_method': 'password',
                    'mfa': 'default',
                }
        ...
        raise AccessDenied()
```

Reply:
Well actually I am thinking to move the block
```py
            if not (credential['type'] == 'password' and credential['password']):
                raise AccessDenied()
``` 
To always apply it even for API keys.

Because you could craft `{'type': 'ldap', 'password': my_api_key}` and currently it would accept it as valid authentication because of that api key fallback, while you asked for another type of authentication.